averageResult changes, header changes (#34)
OpenLiberty/guide-microprofile-rest-client-async: README.adoc

[role='command']
include::{common-includes}/gitclone.adoc[]

> == Redefining client template interface to use asynchronous methods

Reviewer: ```suggestion
== Updating template interface of REST client to use asynchronous methods
```

> == Updating template interface of REST client to use asynchronous methods

Reviewer: ```suggestion
== Updating the template interface of a REST client to use asynchronous methods
```

[role="code_command hotspot", subs="quotes"]
----
#Replace the `JobClient` interface.#
`gateway/src/main/java/io/openliberty/guides/gateway/client/JobClient.java`
----
JobClient.java
[source, Java, linenums, role='code_column hide_tags=copyright']
----
include::finish/gateway/src/main/java/io/openliberty/guides/gateway/client/JobClient.java[]
----

The changes involve changing the return types of the [hotspot=getJobs file=0]`getJobs`, [hotspot=getJob file=0]`getJob`, and [hotspot=createJob file=0]`createJob` methods to return `CompletionStage<T>` types. These changes make the methods asynchronous. Since the methods now have return type of `CompletionStage<T>`, you aren't able to directly manipulate the inner types. As you will see in the next section, you will be able to indirectly use the inner object by chaining callbacks.

> == Updating REST resource to handle HTTP requests asynchronously

Reviewer 2: Is just one REST resource being updated in this example? If so, my suggested update is: 
"Updating the REST resource to asynchronously handle HTTP requests"

Reviewer: In this section, we are only updating one REST resource.  The service itself has multiple resources.  I think having "the" is fine here?  Otherwise, we will have to do something like "Updating REST resource for gateway service ...", which is longer.

Reviewer: Once we close on this, @jonathanxue, we can merge.

Reviewer 2: @yeekangc How about "a" instead of "the"? I think it reads better with an article in front of REST resource. So it would be:
"Updating a REST resource to asynchronously handle HTTP requests"

Reviewer: I went back and forth between "a" and "the" so I am good with it too.  =)

> == Updating the REST resource to asynchronously handle HTTP requests

Reviewer: ```suggestion
== Updating a REST resource to asynchronously handle HTTP requests
```

JAX-RS resources can also have asynchronous methods. So instead of returning a `JobsList` model type, you can return a `CompletionStage<JobsList>` type. Completion stages can be chained together by using the [hotspot=thenApplyAsync file=0]`thenApplyAsync()` method.

[role="code_command hotspot", subs="quotes"]
----

> #Replace the `JobResource` class.#

Reviewer 3: The job resource class for the `gateway` service needs to be renamed, because it's duplicate with the `job/src/main/java/io/openliberty/guides/job/JobResource.java` for the `job` resource.

Reviewer 3: Looks like you had renamed the class in the code to `GatewayJobResource`?

Author: Good catch, this has been corrected now

The changes involve changing the return types of the [hotspot=getJobs file=0]`getJobs`, [hotspot=getJob file=0]`getJob`, and [hotspot=createJob file=0]`createJob` methods to return `CompletionStage<T>` types. These changes make the methods asynchronous. Since the methods now have return type of `CompletionStage<T>`, you aren't able to directly manipulate the inner types. As you will see in the next section, you will be able to indirectly use the inner object by chaining callbacks.

> == Configuring service to respond asynchronously to an HTTP request

Reviewer: ```suggestion
== Updating REST resource to handle HTTP requests asynchronously
```

|Get a specific job |GET \http://localhost:8080/api/jobs/{jobId} |`{"jobId":"661891cb-ad36-4ef4-9bb3-641f973f2964","result":5}`
|===

To create a job, you can use `curl -X POST \http://localhost:8080/api/jobs` command if available on the system. The Postman application can also be used. The request take some time for the job results to return.

The completed jobs JSON output with a created job looks like `{"averageResult":5.0,"count":1,"results":[{"jobId":"661891cb-ad36-4ef4-9bb3-641f973f2964","result":5}]}`. The `averageResult` attribute is the average sleep time of all the jobs. The `count` attribute is the number of jobs, and the `results` attribute contains the list of the jobs. The JSON output for each job has a job ID and a sleep time as the result for the job.

> If no jobs have been created, the JSON output will be this `{"count":0,"results":[]}`. The `count` attribute is 0 because there are no jobs yet, and the `results` attribute is empty for the same reason.

Reviewer 2: Suggested update:
"If no jobs are created, the JSON output is `{"count":0,"results":[]}`. Because there are no jobs yet, the `count` attribute is 0 and the `results` attribute is empty."

|Get a specific job |GET \http://localhost:8080/api/jobs/{jobId} |`{"jobId":"661891cb-ad36-4ef4-9bb3-641f973f2964","result":5}`
|===

> Before any jobs have been created, the JSON output will be this `{"count":0,"results":[]}`. The `count` attribute is 0 because there are no jobs yet, and the `results` attribute is empty for the same reason.

Reviewer 3: I think we want to try fix this issue programatically. Perhaps changing the `OptionalDouble.empty()` to `OptionalDouble.of(0)` in the `JobList.java` class, when there's no jobs being created. Can you try that?

Author: That does work, but it was discussed and in the end was decided to just include a paragraph instead

https://github.com/OpenLiberty/guide-microprofile-rest-client-async/issues/23

Reviewer: The sentence itself needs to be edited.  Also, what we need to call is the difference between the 2 cases (return results)?

Fwiw, I am still on the fence on whether we should fix this.

Reviewer 3: Ah ok, didn't see the discussion in the issue previously. IMO, displaying a 0 average is fine. It eliminates the confusion of having to explain the two different return results.

Reviewer: I wouldn't put this as the first paragraph in the section either.  It will confuse users.  I will put it after the following 2 paragraphs as a clarification and before the concluding paragraph for the section.

[role="code_command hotspot", subs="quotes"]
----

> #Create the `GatewayJobEndpointIT` class.#

Reviewer 3: What's the reason for renaming the test class with "IT"? We haven't been using this convention.
